Add Favorites page

rachelekm opened the pull request:
## Overview

Adds the initial logic and UX for the "Compare" page as well as implementing a refactored limited compare page UI, eg the "Favorites" page.
Includes:
1. Setup redux to add and remove user favorites
2. Hook up compare page menu count to favorites count
3. Add "favorited" to neighborhood preview cards that are in a user's favorite list
4. Hook up favoriting button in neighborhood details to redux logic
5. Add initial "no favorites" compare page UI
6. Add "Favorites" page UI, eg the image carousel from the original "Compare" page design

Note: I implemented commit [4d57e27](https://github.com/azavea/echo-locator/pull/765/commits/4d57e27c105e4484af8d573ca041b222545ea784) to preventing the screen-wide loading spinner to appear on favoriting/un-favoriting. It was appearing as every favorite action was updating user profile, but was enough of a jump in UI that it felt worth addressing. This commit separates a high-level loading spinner for just initial data load and then more component-level loading spinners for data updating loading + user profile loading.

### Checklist

- [x] `fixup!` commits have been squashed
- [x] `CHANGELOG.md` updated with summary of features or fixes, following
      [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) guidelines
- [ ] `README.md` updated if necessary to reflect the changes
- [x] Run `./scripts/format` to lint, format, and fix the application source code.
- [x] CI passes after rebase

### Demo

https://github.com/user-attachments/assets/af9ba19c-de4f-446b-9870-3987c648b121


## Testing Instructions

 * `scripts/server` and login as a user without any destinations set
 * Confirm no regressions logging in and adding initial profile information
 * Confirm on wizard finish, screen-wide loading screen appears as expected
 * Open trips edit modal and edit destinations, on edit confirm loading spinner appears within edit modal
 * Select a neighborhood to open up the details modal. Edit destinations from neighborhood details and, on finish, confirm  loading spinner appears within "Your Trips" section
 * At top of the neighborhood details, select "Favorite" button
 * Confirm no screen-wide loading spinner and button updates to the "unfavorite" version on fulfilled profile update
 * Click the right-hand side of the button group and confirm it navigates to the "favorites" page
 * Confirm the favorites count in the menu header has updated
 * Confirm favorited neighborhood appears as an image card
 * Click the image card and confirm neighborhood details opens as expected
 * Close details and confirm back on favorites page
 * Click the image card's "x" button and confirm favorited neighborhood is removed
 * Confirm the favorites count in the menu header has updated
 * If no favorites, confirm page displays the "no favorites" intro UI to match the figma designs
 * Navigate back to the discover page and begin selecting neighborhoods to favorite
 * As favoriting, confirm their preview cards in both the map and list view are displaying a "favorite" star to match the figma designs
 * Open a favorited neighborhood's details and select the left-hand side "unfavorite" button
 * Confirm neighborhood is removed from user favorites and favorites menu count updates


Resolves #703 #735 #753

rachelekm committed:
Add favorite button logic to details

rachelekm committed:
Prevent UI jump on user profile updating favorites

rachelekm committed:
Add star icon to favorited neighborhood cards

rachelekm committed:
Update menu for favorites

rachelekm committed:
Add base favorites page

rachelekm committed:
Allow favorites to open details + style polish

rachelekm committed:
Fix mobile neighborhood select bounds alignment

rachelekm commented on src/frontend/src/pages/Discover/Map/Map.tsx:
```diff
@@ -85,6 +85,11 @@ const Map = ({ isMobile = true, mapDisplay = true }: Props) => {
+    const SELECT_BOUNDS_PADDING = isMobile
+        ? { top: 100, bottom: 225, left: 100, right: 100 }
+        : 100;
+
```
This adds a little bit more bottom padding to the neighborhood fits bounds on mobile selection/top 10 tour to account for the details preview card display. I didn't have an open issue for this, but I remember it bring brought up in a past RRP demo and felt really noticeable quickly clicking around this time. It ended up being a very quick fix and is a low risk of side effects, so I included it here. If you notice any issues though let's not troubleshoot and just take it out for the sake of budget.

aaronxsu commented on src/frontend/src/components/EditTripsModal.tsx:
```diff
@@ -29,18 +29,25 @@ import { AccordionItem } from "./base/Accordion/AccordionItem";
 
+    // Capture addition/removal of destinations in wizard
+    useEffect(() => {
+        setProfileBuffer(profile);
```
Yes, I'd do the same too, for when there is a buffer

aaronxsu commented on src/frontend/src/components/FavoriteButton.tsx:
```diff
@@ -0,0 +1,78 @@
+import StarIcon from "assets/icons/star.svg?react";
+import Button from "components/base/Button/Button";
+import { useNavigate } from "react-router";
+import { updateUserProfile } from "src/reducers/userProfile/userProfileThunk";
```
Let's remove `src/`. There are other similar instances in other files changed in this PR, I'd recommend cleaning them up too.

rachelekm replied:
🤦‍♀️ sorry this one gets me everytime. Fixed in rebased commits and [9180340](https://github.com/azavea/echo-locator/pull/765/commits/918034051967495e1ffff42017deef9e7244ae5a)

aaronxsu replied:
No worries, I forgot about it before as well. I think we could update the tsconfig and vite config in the future so that auto imports won't add the `src` piece. All good.

rachelekm replied:
That'd be nice! Adding it to my list of incoming tech debt issues 👍

aaronxsu commented on src/frontend/src/components/FavoriteButton.tsx:
```diff
@@ -0,0 +1,78 @@
+        return <></>;
+    }
+    const isFavorited = profile.favorites.includes(zipcode);
+    const profileBuffer = {
```
I think in this case, a buffer might not be strictly necessary, since we are directly updating profile when certain action happens. I find buffers useful when we essentially fill out a form and only persist the form in a final action (i.e. click on save or submit, or exit out of a modal etc). Does it make sense?

rachelekm replied:
That makes sense! I initially copied this over from the user profile wizard as a quick way to have a formatted user profile obj for the update action (since the update request expects `UserProfile` vs `UserProfileSliceState`). I see how keeping this variable name implies a different purpose, so I updated the naming to clarify its simply for formatting not necessarily component-level profile state management.

aaronxsu replied:
Ah, I see what the purpose is now. Thanks for clarifying!

aaronxsu commented on src/frontend/src/pages/Compare/Compare.tsx:
```diff
@@ -0,0 +1,111 @@
+    const neighborhoodDetailsMap = useAppSelector(selectNeighborhoodZipcodeMap);
+    const activeDestination = useAppSelector(selectActiveDestination);
+
+    const profileBuffer = {
```
Similar thought as the above comment regarding the buffer.

rachelekm replied:
Same as above, I added a clarifying comment and renamed this variable

aaronxsu commented on src/frontend/src/pages/Discover/Discover.tsx:
```diff
@@ -111,20 +109,26 @@ const Discover = () => {
+
+    // Separate loading states between loading data and loading data + user profile.
+    // Handling the load state of initial data outside of child components prevents
+    // screen-wide loading spinner during small user profile updates like favoriting.
```
This makes sense to me. I appreciate the comments here.

aaronxsu commented on src/frontend/src/reducers/neighborhoods/neighborhoodsSlice.ts:
```diff
@@ -181,4 +181,22 @@ export const selectNeighborhoodNameByZipcode = createSelector(
         )
 );
 
+export const selectIsFavoriteByZipcode = createSelector(
```
I might be missing something here, but did we end up using this selector?

rachelekm replied:
Ope great catch, I took out this out!

aaronxsu reviewed (approved):
Looking great and working as advertised. I left some suggestions in the comments. Please let me know what you think.

One small thing I'd suggest: after clicking favorite or unfavorite, I'd consider disabling the button when the user profile update is in progress.

rachelekm committed:
Imports cleanup

rachelekm committed:
Update CHANGELOG

rachelekm commented:
Thanks, @aaronxsu! Nice catches that make this much cleaner. I also added a fixup so that the favoriting button is disabled on user profile update. I rebased with develop and added a changelog entry, will merge on successful CI 🎃 

https://github.com/user-attachments/assets/e9369c42-1ad2-4c72-802d-fcea1e21b915

aaronxsu commented:
Thanks for addressing the comments and added that UX enhancement. Looking great to me!
